### ISSUE TITLE
Puma forking multiple cluster under a single worker and misbehaving the configuration

### ISSUE THREAD

**Issue by sriniarul** [puma/puma#2359]
**Describe the bug**
Puma memory utilization increasing overtime. It is forking more clusters under a single worker.
oot     110615 110583 110615  0   27 Sep08 ?        SNl    0:07 puma: cluster worker 0: 110583  
root     110615 110583 110616  0   27 Sep08 ?        SNl    0:01 puma: cluster worker 0: 110583  
root     110615 110583 110631  0   27 Sep08 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583 110653  0   27 Sep08 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583 110742  0   27 Sep08 ?        SNl    4:04 puma: cluster worker 0: 110583  
root     110615 110583 110743  0   27 Sep08 ?        SNl    0:09 puma: cluster worker 0: 110583  
root     110615 110583 110751  0   27 Sep08 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583 110791  0   27 Sep08 ?        SNl    0:01 puma: cluster worker 0: 110583  
root     110615 110583 110922  0   27 Sep08 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583 110923  0   27 Sep08 ?        SNl    0:04 puma: cluster worker 0: 110583  
root     110615 110583 110924  0   27 Sep08 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583 110925  0   27 Sep08 ?        SNl    0:02 puma: cluster worker 0: 110583  
root     110615 110583 111015  0   27 Sep08 ?        SNl    0:01 puma: cluster worker 0: 110583  
root     110615 110583 111016  0   27 Sep08 ?        SNl    1:41 puma: cluster worker 0: 110583  
root     110615 110583 111035  0   27 Sep08 ?        SNl    0:01 puma: cluster worker 0: 110583  
root     110615 110583 111036  0   27 Sep08 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583 111037  0   27 Sep08 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583 111038  0   27 Sep08 ?        SNl    0:08 puma: cluster worker 0: 110583  
root     110615 110583 111039  0   27 Sep08 ?        SNl    0:02 puma: cluster worker 0: 110583  
root     110615 110583 111040  0   27 Sep08 ?        SNl    0:02 puma: cluster worker 0: 110583  
root     110615 110583 111096  0   27 Sep08 ?        SNl    0:06 puma: cluster worker 0: 110583  
root     110615 110583  13438  0   27 08:01 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583  37544  0   27 08:55 ?        SNl    0:22 puma: cluster worker 0: 110583  
root     110615 110583  45122  0   27 09:11 ?        SNl    0:01 puma: cluster worker 0: 110583  
root     110615 110583   7558  0   27 12:13 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583  19425  0   27 12:35 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110615 110583  20362  0   27 12:36 ?        SNl    0:00 puma: cluster worker 0: 110583  
root     110617 110583 110617  0   34 Sep08 ?        SNl    0:07 puma: cluster worker 1: 110583  
root     110617 110583 110618  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 1: 110583  
root     110617 110583 110636  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 110652  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 110787  0   34 Sep08 ?        SNl    4:20 puma: cluster worker 1: 110583  
root     110617 110583 110788  0   34 Sep08 ?        SNl    0:09 puma: cluster worker 1: 110583  
root     110617 110583 110802  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 110859  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 1: 110583  
root     110617 110583 110932  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 110933  0   34 Sep08 ?        SNl    0:04 puma: cluster worker 1: 110583  
root     110617 110583 110934  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 110935  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 1: 110583  
root     110617 110583 111067  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 1: 110583  
root     110617 110583 111068  0   34 Sep08 ?        SNl    1:41 puma: cluster worker 1: 110583  
root     110617 110583 111080  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 111081  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 111082  0   34 Sep08 ?        SNl    0:06 puma: cluster worker 1: 110583  
root     110617 110583 111083  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 1: 110583  
root     110617 110583 111084  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 1: 110583  
root     110617 110583 111086  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 1: 110583  
root     110617 110583 111103  0   34 Sep08 ?        SNl    0:06 puma: cluster worker 1: 110583  
root     110617 110583  47084  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  47085  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  47086  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  47087  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  47088  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  47089  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  13228  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  93605  0   34 10:49 ?        SNl    0:13 puma: cluster worker 1: 110583  
root     110617 110583  94649  0   34 10:51 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583 100622  0   34 11:03 ?        SNl    0:11 puma: cluster worker 1: 110583  
root     110617 110583 126935  0   34 11:51 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110617 110583  18727  0   34 12:34 ?        SNl    0:01 puma: cluster worker 1: 110583  
root     110617 110583  20518  1   34 12:37 ?        SNl    0:00 puma: cluster worker 1: 110583  
root     110619 110583 110619  0   29 Sep08 ?        SNl    0:08 puma: cluster worker 2: 110583  
root     110619 110583 110620  0   29 Sep08 ?        SNl    0:01 puma: cluster worker 2: 110583  
root     110619 110583 110641  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583 110658  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583 110835  0   29 Sep08 ?        SNl    4:18 puma: cluster worker 2: 110583  
root     110619 110583 110836  0   29 Sep08 ?        SNl    0:09 puma: cluster worker 2: 110583  
root     110619 110583 110852  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583 110897  0   29 Sep08 ?        SNl    0:01 puma: cluster worker 2: 110583  
root     110619 110583 110937  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583 110938  0   29 Sep08 ?        SNl    0:04 puma: cluster worker 2: 110583  
root     110619 110583 110939  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583 110940  0   29 Sep08 ?        SNl    0:02 puma: cluster worker 2: 110583  
root     110619 110583 111242  0   29 Sep08 ?        SNl    0:01 puma: cluster worker 2: 110583  
root     110619 110583 111243  0   29 Sep08 ?        SNl    1:38 puma: cluster worker 2: 110583  
root     110619 110583 111244  0   29 Sep08 ?        SNl    0:01 puma: cluster worker 2: 110583  
root     110619 110583 111245  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583 111246  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583 111247  0   29 Sep08 ?        SNl    0:06 puma: cluster worker 2: 110583  
root     110619 110583 111248  0   29 Sep08 ?        SNl    0:02 puma: cluster worker 2: 110583  
root     110619 110583 111249  0   29 Sep08 ?        SNl    0:02 puma: cluster worker 2: 110583  
root     110619 110583 111285  0   29 Sep08 ?        SNl    0:06 puma: cluster worker 2: 110583  
root     110619 110583  19161  0   29 Sep08 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583  77695  0   29 10:19 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583  77696  0   29 10:19 ?        SNl    0:01 puma: cluster worker 2: 110583  
root     110619 110583   5124  0   29 12:08 ?        SNl    0:05 puma: cluster worker 2: 110583  
root     110619 110583  13473  0   29 12:24 ?        SNl    0:02 puma: cluster worker 2: 110583  
root     110619 110583  18536  1   29 12:33 ?        SNl    0:02 puma: cluster worker 2: 110583  
root     110619 110583  20457  0   29 12:37 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110619 110583  20531  2   29 12:37 ?        SNl    0:00 puma: cluster worker 2: 110583  
root     110621 110583 110621  0   28 Sep08 ?        SNl    0:07 puma: cluster worker 3: 110583  
root     110621 110583 110622  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 3: 110583  
root     110621 110583 110632  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583 110654  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583 110768  0   28 Sep08 ?        SNl    4:52 puma: cluster worker 3: 110583  
root     110621 110583 110769  0   28 Sep08 ?        SNl    0:09 puma: cluster worker 3: 110583  
root     110621 110583 110777  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583 110830  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 3: 110583  
root     110621 110583 110927  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583 110928  0   28 Sep08 ?        SNl    0:04 puma: cluster worker 3: 110583  
root     110621 110583 110929  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583 110930  0   28 Sep08 ?        SNl    0:02 puma: cluster worker 3: 110583  
root     110621 110583 110996  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 3: 110583  
root     110621 110583 110997  0   28 Sep08 ?        SNl    1:39 puma: cluster worker 3: 110583  
root     110621 110583 111000  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 3: 110583  
root     110621 110583 111009  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583 111010  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583 111011  0   28 Sep08 ?        SNl    0:05 puma: cluster worker 3: 110583  
root     110621 110583 111012  0   28 Sep08 ?        SNl    0:02 puma: cluster worker 3: 110583  
root     110621 110583 111013  0   28 Sep08 ?        SNl    0:02 puma: cluster worker 3: 110583  
root     110621 110583 111092  0   28 Sep08 ?        SNl    0:07 puma: cluster worker 3: 110583  
root     110621 110583  13445  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583  89282  0   28 10:41 ?        SNl    0:01 puma: cluster worker 3: 110583  
root     110621 110583  89517  0   28 10:42 ?        SNl    0:01 puma: cluster worker 3: 110583  
root     110621 110583 116972  0   28 11:32 ?        SNl    0:10 puma: cluster worker 3: 110583  
root     110621 110583  12006  0   28 12:21 ?        SNl    0:02 puma: cluster worker 3: 110583  
root     110621 110583  20054  0   28 12:36 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110621 110583  20078  0   28 12:36 ?        SNl    0:00 puma: cluster worker 3: 110583  
root     110623 110583 110623  0   28 Sep08 ?        SNl    0:08 puma: cluster worker 4: 110583  
root     110623 110583 110624  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 4: 110583  
root     110623 110583 110635  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 110651  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 110691  0   28 Sep08 ?        SNl    2:41 puma: cluster worker 4: 110583  
root     110623 110583 110692  0   28 Sep08 ?        SNl    0:09 puma: cluster worker 4: 110583  
root     110623 110583 110700  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 110727  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 4: 110583  
root     110623 110583 110902  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 110903  0   28 Sep08 ?        SNl    0:04 puma: cluster worker 4: 110583  
root     110623 110583 110904  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 110905  0   28 Sep08 ?        SNl    0:02 puma: cluster worker 4: 110583  
root     110623 110583 110908  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 4: 110583  
root     110623 110583 110909  0   28 Sep08 ?        SNl    1:41 puma: cluster worker 4: 110583  
root     110623 110583 110915  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 110916  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 110917  0   28 Sep08 ?        SNl    0:09 puma: cluster worker 4: 110583  
root     110623 110583 110918  0   28 Sep08 ?        SNl    0:02 puma: cluster worker 4: 110583  
root     110623 110583 110919  0   28 Sep08 ?        SNl    0:02 puma: cluster worker 4: 110583  
root     110623 110583 110966  0   28 Sep08 ?        SNl    0:01 puma: cluster worker 4: 110583  
root     110623 110583 110972  0   28 Sep08 ?        SNl    0:07 puma: cluster worker 4: 110583  
root     110623 110583  76061  0   28 Sep08 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583 109616  0   28 11:19 ?        SNl    0:05 puma: cluster worker 4: 110583  
root     110623 110583  12897  0   28 12:23 ?        SNl    0:01 puma: cluster worker 4: 110583  
root     110623 110583  15979  1   28 12:28 ?        SNl    0:07 puma: cluster worker 4: 110583  
root     110623 110583  16011  0   28 12:28 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583  16015  0   28 12:28 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110623 110583  20507  0   28 12:37 ?        SNl    0:00 puma: cluster worker 4: 110583  
root     110625 110583 110625  0   35 Sep08 ?        SNl    0:07 puma: cluster worker 5: 110583  
root     110625 110583 110626  0   35 Sep08 ?        SNl    0:01 puma: cluster worker 5: 110583  
root     110625 110583 110634  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583 110656  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583 110819  0   35 Sep08 ?        SNl    4:02 puma: cluster worker 5: 110583  
root     110625 110583 110820  0   35 Sep08 ?        SNl    0:09 puma: cluster worker 5: 110583  
root     110625 110583 110833  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583 110896  0   35 Sep08 ?        SNl    0:01 puma: cluster worker 5: 110583  
root     110625 110583 110942  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583 110943  0   35 Sep08 ?        SNl    0:04 puma: cluster worker 5: 110583  
root     110625 110583 110944  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583 110945  0   35 Sep08 ?        SNl    0:02 puma: cluster worker 5: 110583  
root     110625 110583 111250  0   35 Sep08 ?        SNl    0:01 puma: cluster worker 5: 110583  
root     110625 110583 111251  0   35 Sep08 ?        SNl    1:40 puma: cluster worker 5: 110583  
root     110625 110583 111253  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583 111254  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583 111255  0   35 Sep08 ?        SNl    0:06 puma: cluster worker 5: 110583  
root     110625 110583 111256  0   35 Sep08 ?        SNl    0:02 puma: cluster worker 5: 110583  
root     110625 110583 111257  0   35 Sep08 ?        SNl    0:02 puma: cluster worker 5: 110583  
root     110625 110583 111259  0   35 Sep08 ?        SNl    0:01 puma: cluster worker 5: 110583  
root     110625 110583 111288  0   35 Sep08 ?        SNl    0:07 puma: cluster worker 5: 110583  
root     110625 110583  18237  0   35 Sep08 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583  46074  0   35 Sep08 ?        SNl    0:01 puma: cluster worker 5: 110583  
root     110625 110583 117459  0   35 11:33 ?        SNl    0:07 puma: cluster worker 5: 110583  
root     110625 110583   1122  0   35 12:01 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583   1123  0   35 12:01 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583   1124  0   35 12:01 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583   1125  0   35 12:01 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583   1126  0   35 12:01 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583   1127  0   35 12:01 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583  17878  0   35 12:32 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583  17990  0   35 12:32 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583  18083  0   35 12:32 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583  19732  0   35 12:35 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110625 110583  20496  1   35 12:37 ?        SNl    0:00 puma: cluster worker 5: 110583  
root     110627 110583 110627  0   34 Sep08 ?        SNl    0:07 puma: cluster worker 6: 110583  
root     110627 110583 110628  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 6: 110583  
root     110627 110583 110633  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 110650  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 110716  0   34 Sep08 ?        SNl    4:35 puma: cluster worker 6: 110583  
root     110627 110583 110717  0   34 Sep08 ?        SNl    0:09 puma: cluster worker 6: 110583  
root     110627 110583 110725  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 110764  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 6: 110583  
root     110627 110583 110911  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 110912  0   34 Sep08 ?        SNl    0:04 puma: cluster worker 6: 110583  
root     110627 110583 110913  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 110914  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 6: 110583  
root     110627 110583 111347  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 6: 110583  
root     110627 110583 111348  0   34 Sep08 ?        SNl    1:40 puma: cluster worker 6: 110583  
root     110627 110583 111369  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 111370  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 111371  0   34 Sep08 ?        SNl    0:05 puma: cluster worker 6: 110583  
root     110627 110583 111372  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 6: 110583  
root     110627 110583 111373  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 6: 110583  
root     110627 110583 111375  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 6: 110583  
root     110627 110583 111393  0   34 Sep08 ?        SNl    0:06 puma: cluster worker 6: 110583  
root     110627 110583  28838  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583  28839  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583  28840  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583  28841  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583  28842  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583  28843  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583  91353  0   34 Sep08 ?        SNl    0:03 puma: cluster worker 6: 110583  
root     110627 110583  49566  0   34 09:21 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583 113933  0   34 11:27 ?        SNl    0:12 puma: cluster worker 6: 110583  
root     110627 110583   3870  0   34 12:05 ?        SNl    0:06 puma: cluster worker 6: 110583  
root     110627 110583  15022  0   34 12:26 ?        SNl    0:05 puma: cluster worker 6: 110583  
root     110627 110583  16652  0   34 12:29 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110627 110583  16820  0   34 12:30 ?        SNl    0:00 puma: cluster worker 6: 110583  
root     110629 110583 110629  0   34 Sep08 ?        SNl    0:08 puma: cluster worker 7: 110583  
root     110629 110583 110630  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 7: 110583  
root     110629 110583 110640  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 110657  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 110856  0   34 Sep08 ?        SNl    5:04 puma: cluster worker 7: 110583  
root     110629 110583 110857  0   34 Sep08 ?        SNl    0:09 puma: cluster worker 7: 110583  
root     110629 110583 110882  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 110889  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 7: 110583  
root     110629 110583 110898  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 7: 110583  
root     110629 110583 110947  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 110948  0   34 Sep08 ?        SNl    0:04 puma: cluster worker 7: 110583  
root     110629 110583 110949  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 110950  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 7: 110583  
root     110629 110583 110979  0   34 Sep08 ?        SNl    1:38 puma: cluster worker 7: 110583  
root     110629 110583 110987  0   34 Sep08 ?        SNl    0:01 puma: cluster worker 7: 110583  
root     110629 110583 110991  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 110992  0   34 Sep08 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 110993  0   34 Sep08 ?        SNl    0:03 puma: cluster worker 7: 110583  
root     110629 110583 110994  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 7: 110583  
root     110629 110583 110995  0   34 Sep08 ?        SNl    0:02 puma: cluster worker 7: 110583  
root     110629 110583 111089  0   34 Sep08 ?        SNl    0:05 puma: cluster worker 7: 110583  
root     110629 110583  75943  0   34 Sep08 ?        SNl    0:03 puma: cluster worker 7: 110583  
root     110629 110583  87115  0   34 05:45 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583  87116  0   34 05:45 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583  87117  0   34 05:45 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583  87118  0   34 05:45 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583  87119  0   34 05:45 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583  87120  0   34 05:45 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583  92976  0   34 06:00 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 127088  0   34 11:51 ?        SNl    0:07 puma: cluster worker 7: 110583  
root     110629 110583 129617  0   34 11:56 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583 129664  0   34 11:56 ?        SNl    0:00 puma: cluster worker 7: 110583  
root     110629 110583  14606  0   34 12:26 ?        SNl    0:02 puma: cluster worker 7: 110583  
root     110629 110583  16823  1   34 12:30 ?        SNl    0:07 puma: cluster worker 7: 110583  

when checking number of clusters under each worker with this cmd : ps -o nlwp $(pgrep puma)
NLWP
   5
  28
  35
  28
  28
  28
  34
  35
  35

**Puma config:**
workers 8
# Min and Max threads per worker
threads 1, 5
app_dir = File.expand_path("../..", __FILE__)
shared_dir = "#{app_dir}/shared"

# Default to production
rails_env = ENV['RAILS_ENV'] || "production"
environment rails_env

# Set up socket location
bind "unix://#{shared_dir}/sockets/puma.sock"
#port        ENV['PORT']     || 3000
# Logging
stdout_redirect "#{shared_dir}/log/puma.stdout.log", "#{shared_dir}/log/puma.stderr.log", true
# Set master PID and state locations
pidfile "#{shared_dir}/pids/puma.pid"

state_path "#{shared_dir}/pids/puma.state"

activate_control_app

on_worker_boot do
require "active_record"
  ActiveRecord::Base.connection.disconnect! rescue ActiveRecord::ConnectionNotEstablished
  ActiveRecord::Base.establish_connection(YAML.load_file("#{app_dir}/config/database.yml")[rails_env])
end

**To Reproduce**
unable to reproduce this as it normal during restarting and memory consumption goes high. While checking the process the number of clusters under each worker is also high.

**Expected behavior**
Each worker should fork limited cluster as defined in the configuration file.

**Desktop (please complete the following information):**
 - OS: [RHEL-7.*]
 - Puma Version [e.g. 4.3.5]

**Comment by nateberkopec** [comment]
> Each worker should fork limited cluster as defined in the configuration file.

I don't understand. The output you pasted here has no labels, so I can't follow what's going on.

Can you reproduce the issue with a simple/blank Rack application?

If you can't, I'm guessing you have code _in your application_ which is calling `fork`.

**Comment by sriniarul** [comment]
From past one month, I didn't got any issue. But this happened recently. I just need to figure out the cause. puma/code.
Can you please clarify this doubt.
if a thread get locked does puma create another thread under that worker process?
I am using mutex in one of my service.

**Comment by nateberkopec** [comment]
> if a thread get locked does puma create another thread under that worker process

I'm not sure what you mean. If your Puma thread configuration is set to something like minimum 0 threads, maximum 5 threads, then if 1 thread starts processing a request, a second thread will get created if the first thread releases the GVL and there is another request waiting on the socket.

**Comment by sriniarul** [comment]
Here is what I found. It's not the issue of puma. It sounds like ruby GC is not collecting the unused objects. Our application mostly does big DB queries. And while responding the request it is cleating tons of objects and it remains in the memory. The htop cmd on linux shows duplicate process names, that makes me doubt over the puma process. But after starting puma in only worker mode I noticed it and found that each worker memory usage consistently increasing over time. Then done some research over the garbage collector and found the root cause. I am closing the issue. Thanks for your time.
